### ISSUE TITLE
catchup=fs: allow slash symbol in stream name

### ISSUE THREAD

**Issue by door** [kodi-pvr/pvr.iptvsimple#466]
Hi,

Stream name in flussonic can contain "/":
```
#EXTINF:-1 catchup="fs",Channel G
http://list.tv:8888/stream/name/with/shashes/mono.m3u8?token=secret
```
Here stream name is `stream/name/with/shashes`.

The regex in `GenerateFlussonicCatchupSource` should be extended to accept such names.

**Comment by wsnipex** [comment]
no, slashes are reserverd chars in http, so "stream/name/with/shashes" is not a valid filename.
fix: url encode invalid chars in the name

**Comment by door** [comment]
Flussonic stream name not equals to file or directory name. Also middleware may interpret stream name and give redirect to real name.

Where from comes this limitation, that stream name has to be valid filename? HTTP does not impose any restrictions here.

**Comment by wsnipex** [comment]
HTTP has no concept of "flussonic stream names" (whatever that is), it only knows protocol, path, filename, query string components.
And according to the relevent RFCs, a filename may not contain reserved characters.

Maybe start with describing your issue in detail, complete with a debug log

**Comment by door** [comment]
Afaik, there is no "filename" in URI, only path.

This is flussonic sample config:
```
stream a/b/c {
   url fake://fake;
}
```

This is m3u playlist:
```
#EXTINF:-1 catchup="fs",Channel G
http://list.tv:8888/a/b/c/mono.m3u8?token=secret
```

The "a/b/c" part is the stream name, and archive links will look like:
`http://list.tv:8888/a/b/c/timeshift_rel-3600.m3u8?token=secret`
`http://list.tv:8888/a/b/c/video-1610452406-3600.m3u8`
and so on.

**Comment by door** [comment]
Kodi installed from http://ppa.launchpad.net/team-xbmc/xbmc-nightly/ubuntu repo:
```
kodi 2:19.0+git20210101.0302-3384132053-0~focal
kodi-pvr-iptvsimple 7.1.0-1~focal
```

fs.m3u:
```
#EXTM3U
#EXTINF:-1 catchup="fs",s1/s2
http://flussonic:8080/s1/s2/video.m3u8
#EXTINF:-1 catchup="fs",s3
http://flussonic:8080/s3/video.m3u8
```

s1/s2 and s3 streams both play live.

But searching for "Flussonic" in kodi.log gives only urls for s3:
```
Flussonic - s3: http://flussonic:8080/s3/video-timeshift_rel-{offset:1}.m3u8
```
and no s1/s2.
[kodi.log](https://github.com/kodi-pvr/pvr.iptvsimple/files/5802351/kodi.log)

**Comment by phunkyfish** [comment]
To take one of your examples:

http://list.tv:8888/a/b/c/mono.m3u8?token=secret

Should be:

http://list.tv:8888/a%2fb%2fc/mono.m3u8?token=secret

This should still work and also be a valid HTTP url.

Similarly: http://flussonic:8080/s1/s2/video.m3u8
should be: http://flussonic:8080/s1%2fs2/video.m3u8

**Comment by door** [comment]
I updated the playlist:
```
#EXTM3U
#EXTINF:-1 catchup="fs",s1/s2
http://flussonic:8080/s1%2fs2/video.m3u8
#EXTINF:-1 catchup="fs",s3
http://flussonic:8080/s3/video.m3u8
```

and s1/s2 stream is not playing anymore.

```
curl http://flussonic:8080/s1%2fs2/video.m3u8
Not found
```

```
curl http://flussonic:8080/s1/s2/video.m3u8
#EXTM3U
#EXT-X-TARGETDURATION:5
...
```

**Comment by phunkyfish** [comment]
The only other I can see of allowing this is to change the regex which builds the catchup URL.

Try these test builds: https://jenkins.kodi.tv/blue/organizations/jenkins/kodi-pvr%2Fpvr.iptvsimple/detail/PR-502/14/artifacts

**Comment by phunkyfish** [comment]
@door did you get a chance to test this?
